hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/utils/HddsVolumeUtil.java

```diff
@@ -234,4 +257,154 @@ public static boolean checkVolume(HddsVolume hddsVolume, String scmId,
+    DbVolume chosenDbVolume = null;
+    File dbParentDir;
+
+    if (dbVolumeSet == null || dbVolumeSet.getVolumesList().isEmpty()) {
+      // No extra db volumes specified, just create db under the HddsVolume.
+      dbParentDir = new File(hddsVolume.getStorageDir(), clusterID);
+    } else {
+      // Randomly choose a DbVolume for simplicity.
+      List<DbVolume> dbVolumeList = StorageVolumeUtil.getDbVolumesList(
+          dbVolumeSet.getVolumesList());
+      chosenDbVolume = dbVolumeList.get(
+          ThreadLocalRandom.current().nextInt(dbVolumeList.size()));
+      dbParentDir = new File(chosenDbVolume.getStorageDir(), clusterID);
+    }
+
+    // Init subdir with the storageID of HddsVolume.
+    File storageIdDir = new File(dbParentDir, hddsVolume.getStorageID());
+    if (!storageIdDir.mkdirs() && !storageIdDir.exists()) {
+      throw new IOException("Can't make subdir under "
+          + dbParentDir.getAbsolutePath() + " for volume "
+          + hddsVolume.getStorageID());
+    }
+
+    // Init the db instance for HddsVolume under the subdir above.
+    String containerDBPath = new File(storageIdDir, CONTAINER_DB_NAME)
+        .getAbsolutePath();
+    try {
+      initPerDiskDBStore(containerDBPath, conf);
+    } catch (IOException e) {
+      throw new IOException("Can't init db instance under path "
+          + containerDBPath + " for volume " + hddsVolume.getStorageID());
+    }
+
+    // Set the dbVolume and dbParentDir of the HddsVolume for db path lookup.
+    hddsVolume.setDbVolume(chosenDbVolume);
+    hddsVolume.setDbParentDir(storageIdDir);
+  }
+
+  /**
+   * Load already formatted db instances for all HddsVolumes.
+   * @param hddsVolumeSet
+   * @param dbVolumeSet
+   * @param conf
+   * @param logger
+   */
+  public static void loadAllHddsVolumeDbStore(MutableVolumeSet hddsVolumeSet,
+      MutableVolumeSet dbVolumeSet, ConfigurationSource conf, Logger logger) {
+    // Scan subdirs under the db volumes and build a one-to-one map
+    // between each HddsVolume -> DbVolume.
+    mapDbVolumesToDataVolumesIfNeeded(hddsVolumeSet, dbVolumeSet);
+
+    List<HddsVolume> hddsVolumeList = StorageVolumeUtil.getHddsVolumesList(
+        hddsVolumeSet.getVolumesList());
+
+    for (HddsVolume volume : hddsVolumeList) {
+      String clusterID = volume.getClusterID();
+
+      // DN startup for the first time, not registered yet,
+      // so the DbVolume is not formatted.
+      if (clusterID == null) {
+        continue;
+      }
+
+      File clusterIdDir = new File(volume.getDbVolume() == null ?
+          volume.getStorageDir() : volume.getDbVolume().getStorageDir(),
+          clusterID);
+
+      if (!clusterIdDir.exists()) {
+        if (logger != null) {
+          logger.error("HddsVolume {} db instance not formatted, " +
+                  "clusterID {} directory not found",
+              volume.getStorageDir().getAbsolutePath(),
+              clusterIdDir.getAbsolutePath());
+        }
+        continue;
```

Comment:
Here in this function, we are iterating all of the db instances for all `HddsVolume`s, so we don't just throw and bailout.
We should go on trying to load as many db instances as possible.

Reply:
For storage, data integrity and security is very important.  So we better be conservative and fail the code fast if something happens which requires admin's notice and interference. So please throw out the exception here.

Reply:
Just a simple question here: say we have 100 disks for a DN, on extra SSDs, so db instances are on the same disk as data. If a disk failure causes the db init fails, do we continue for other disks or just throw and out?
I don't think we will stop DN startup on a disk failure for now since disk failure is common, it does not really hurt data integrity or security for a distributed system.

The failedDbVolumeTolerated defaults to -1 means we'll have at least one dbVolume if configured any, please check `hasEnoughVolumes#hasEnoughVolumes`.

Reply:
It depends on how many disk failure is tolerated.  For example, in HDFS,  default disk failure tolerated is 0, which means as long as one disk fails, DN cannot start up.  Admin may change the configuration based on his/her estimation of disk failure rate.  That's acceptable and the risk is on Admin.  So "at least one disk volume" is not acceptable, the risk is too high. 

Think about this case, all HDDS volumes function well while one of the two configured dbVolumes is down at DN startup.  How will the impacted HDDS volume behave? Create a new RocksDB instance on the remaining dbVolume and then go on to provide service?  I think we should persist this HDDS volume to RocksDB instance relation on HDDS
version file.

Reply:
1. For the failed volume tolerated value, read this thread https://github.com/apache/ozone/pull/2243 which you've participant in, there's a discussion about the default 0 or -1. You could send a patch to fix it overall with the above reasons later after you have decided about the best default value, but not in this thread.
2. For the "throw or continue" problem, first please provide an answer to my question first, what is the desired behavior(with no extra ssds) in your mind and the reason, then we try to reach an agreement.
3. For the new question you raised, suppose there are 2 DbVolumes and one of them is down, the db instances on that one is unreadable and all related HddsVolumes are unreadable. This scenario equals to that half of the HddsVolumes are down. This is not a very extreme case, even the whole DN is down, data integrity and security is not fatally harmed. But there won't be any new db instances created on the remaining dbVolume as you said since db instances are only created when we try to format an HddsVolume on the DN registration, please recheck.

Reply:
For the HddsVolume -> Instance relation, it is mapped each time on DN startup, we scan each `DbVolume` trying to find the `StorageID` of each `HddsVolume` under it and then we have the map rebuilt, so no need to persist in VERSION file.
The container replication logic will be introduced in later patches, and containers can be replicated as they are currently. If you are providing an extra command util, that's good.

Reply:
> 1. For the failed volume tolerated value, read this thread [HDDS-5219. Limit number of bad volumes by dfs.datanode.failed.volumes.tolerated #2243](https://github.com/apache/ozone/pull/2243) which you've participant in, there's a discussion about the default 0 or -1. You could send a patch to fix it overall with the above reasons later after you have decided about the best default value, but not in this thread.

It can be done in another PR. I will take care of that. 

> 
>     2. For the "throw or continue" problem, first please provide an answer to my question first, what is the desired behavior(with no extra ssds) in your mind and the reason, then we try to reach an agreement.

If rocksdb is on the data volume. We follow the data volume rule. If the rocksdb load is failed,  it means this data volume fails.  If the failed data volume exceeds the tolerance,  then DN fails.  Is this reasonable? 

> 
>     3. For the new question you raised, suppose there are 2 DbVolumes and one of them is down, the db instances on that one is unreadable and all related HddsVolumes are unreadable. This scenario equals to that half of the HddsVolumes are down. This is not a very extreme case, even the whole DN is down, data integrity and security is not fatally harmed. But there won't be any new db instances created on the remaining dbVolume as you said since db instances are only created when we try to format an HddsVolume on the DN registration, please recheck.

If I remember correctly, DN will register to SCM on every startup, which means if one DbVolume failed, all impacted HddsVolume will have no DbVolume and ParentStorageDir set, then it need create a new RocksDB in function #volumeCheck. Right?

Reply:
1. Thanks for it, I'll help review it then if able to.
2. The behavior you stated sounds reasonable to me(I'll check bad volume if db load failed in the next push), so as I understand, we should continue rather than throw&out when we hit a bad volume. Agreed?
3. Let me clarify it a bit. A DbVolume is only a place holder for potential db instances for HddsVolumes, it is formatted (only create a root directory and VERSION file)when DN register happens, before all HddsVolumes. A db instance is formatted for each HddsVolume when `checkVolume` is called for the first time(after all DbVolumes are formatted). Check my comment below about the condition check `hddsFiles.length == 1`. Upon the first register of DN -> SCM, DN get a clusterID back and create a directory(e.g. /data1/hdds/CID-<clusterID>) for each HddsVolume, we only create a db instance together with the clusterID directory. On the 2nd, 3rd,... DN registers, `checkVolume` will found that `hddsFiles.length == 2`, so no redundant db instance will be created. Does it make sense?

Reply:
>     2. The behavior you stated sounds reasonable to me(I'll check bad volume if db load failed in the next push), so as I understand, we should continue rather than throw&out when we hit a bad volume. Agreed?

After all the discussions, we should add the failed hddsVolume to failed volume list, then check that if failed volumes exceed the limit, right?  

> 
>     3. Let me clarify it a bit. A DbVolume is only a place holder for potential db instances for HddsVolumes, it is formatted when DN register happens(only for now, I'll change it to format on DN startup), before all HddsVolumes. A db instance is formatted for each HddsVolume when `checkVolume` is called for the first time(after all DbVolumes are formatted). Check my comment below about the condition check `hddsFiles.length == 1`. Upon the first register of DN -> SCM, DN get a clusterID back and create a directory(e.g. /data1/hdds/CID-) for each HddsVolume, we only create a db instance together with the clusterID directory. On the 2nd, 3rd,... DN restart and registers, `checkVolume` will found that `hddsFiles.length == 2`, so no redundant db instance will be created. Does it make sense?

Here are two cases, 
1. RocksDB on HddsVolume.  Your explanation makes sense. But, current hddsFiles.length == 2 logic is SCM HA upgrade fix.  If you depends on hddsFiles.length 2 to detect whether RocksDB created or not, you need to distinguish these two different cases. 
2. RocksDB on dbVolume. In the case, RocksDB existence is irrelevant with hddsFiles.length, right?

Reply:
For 2, failed volume check doesn't belong here, it is an async check and DbVolumes are handled as other volumes, please check `StorageVolumeChecker.java`.  And I still want a clear choice between (A)"throw & out" (B)"continue"?

For 3.1, let me explain `hddsFiles.length == 2`:  one file is for VERSION; the other file is either clusterIDDir or scmIDDir, SCM HA checks whether the Dir is (A)clusterIDDir or (B)scmIDDir. If A, do nothing because it means SCM HA is finalized; If B, upgrade this volume for SCM HA by linking. But no matter A or B, we have an IDDir which implies that this HddsVolume is formatted on the first DN register, and we won't do db format again for it neither on DN register nor DN restart. Please check the code inside this function for the words above `VersionedDatanodeFeatures.ScmHA.upgradeVolumeIfNeeded`.
So here I mean I don't actually depends on SCM HA to check whether we have to format db instance or not, clear?

For 3.2, after 3.1, I think you should know about the meaning of `hddsFiles.length`. And let me say more about when we have the chance to create a db instance:
- Along with HddsVolume format on DN first register, so only when we have`hddsFiles.length == 1` which means that we only got a VERSION file on the HddsVolume, so we have to create the clusterIDDir/scmIDDir only on the first DN register.
- Along with load on DN restart, I think this is where you actually suspect that we may have duplicate db instance creation. But I should say that, there is a check `!storageIdDir.exists()` at line 350 of file `HddsVolumeUtil.java` that checks if the subdirectry with `StorageID` exists, we only load the db instance when this exists. And for now I think I should have a stricter check for the `container.db` directory exactly, then we prevent from creating db instances on all pathes. Make sense?

Reply:
And if you have time, please check the descriptions in the JIRA above, there you could see that layout picture of db locations, and you'll be more clear about this.

Reply:
@guihecheng , thanks for the explanation. Got the logic of  HddsFiles.length now. Just feel that the logic is not that straightforward and a little hard to understand.  So in a summary, we only create the rocksdb on a new DN's first time registration. After that, the only place where DN will load rocksDB instances is in #loadAllHddsVolumeDbStore. 

Let's back to the use case that one of two dbVolumes is failed. In #mapDbVolumesToDataVolumesIfNeeded, say half of the HddsVolumes have their dbVolume loaded and set, half of them don't because of their corresponding dbVolume is failed. We need to identify this case in #mapDbVolumesToDataVolumesIfNeeded and present the impacted HddsVolume from creating a new RocksDB under its sub-directory.

Reply:
OK, then I think I can take (B) for the answer to my question and keep the "continue" way if there is a db load failure now if you are not against it. Of course I'll rewrite the confusing `loadAllHddsVolumeDbStore ` into a cleaner function like:
`
void loadAllHddsVolumeDbStore() {
    try {
       loadHddsVolumeDbStore();
    } catch (IOException e) {
        LOG.error(...);
        continue;
    }
}
`
In this way, I think we'll are be satisfied, there is Exception thrown and we'll continue the load for other good volumes.

Reply:
At last, I think I should really add dedicated tests for the cases you raised seriously, thanks for checking it.

Reply:
Another case I just thought about is upgrading. If we enable this feature on an existing cluster,  HddsFiles.length will be 2 in such case, so which piece of code will help us to create this rocksDB per disk instance?

Reply:
We'll have a dedicated patch for non-rolling upgrade at last and we could define non-rolling upgrade hooks to help create the db instances upon `FinalizeUpgrade`.
We could implement those hooks by referring to the hooks defined for SCM HA.
(BTW, in the internal version we disabled the non-rolling upgrade feature, so the db instance logic is a bit different from the code present here.)

Reply:
OK,  sounds good.